src/std/net/s3/api.ss

```diff
@@ -0,0 +1,242 @@
+;;; -*- Gerbil -*-
+;;; (C) vyzo
+;;; AWS S3 client
+(import "sigv4"
+        :std/net/request
+        :std/misc/func
+        :std/contract
+        :std/net/uri
+        :std/crypto/digest
+        :std/text/hex
+        :std/xml
+        :std/error
+        :std/sugar
+        :std/srfi/19)
+(export (struct-out s3-client bucket) S3ClientError)
+
+; precomputed empty sha256
+(def emptySHA256 #u8(227 176 196 66 152 252 28 20 154 251 244 200 153 111 185
+                     36 39 174 65 228 100 155 147 76 164 149 153 27 120 82 184 85))
+
+(defstruct s3-client (endpoint access-key secret-key region)
+  final: #t
+  constructor: :init!)
+
+(defstruct bucket (client name region)
+  final: #t)
+
+(deferror-class (S3ClientError Error) () s3-client-error?)
+
+(defraise/context (raise-s3-error where message irritants ...)
+  (S3ClientError message irritants: [irritants ...]))
+
+; Initializes a `s3-client`. Primarily responsible for holding onto credentials
+(defmethod {:init! s3-client}
+  (lambda (self
+            (endpoint "s3.amazonaws.com")
+            (access-key (getenv "AWS_ACCESS_KEY_ID" #f))
+            (secret-key (getenv "AWS_SECRET_ACCESS_KEY" #f))
+            (region (getenv "AWS_DEFAULT_REGION" "us-east-1")))
+    (using (self :- s3-client)
+      (set! self.endpoint endpoint)
+      (set! self.access-key access-key)
+      (set! self.secret-key secret-key)
+      (set! self.region region))))
+
+; Retrieves buckets accessible to this client.
+(defmethod {list-buckets s3-client} ; => (list : bucket)
+  (lambda (self)
+    (using (self :- s3-client)
+           (let* ((req (s3-request/error self verb: 'GET))
+                  (xml (s3-parse-xml req))
+                  (buckets (sxml-find xml (sxml-e? 's3:Buckets) sxml-children))
+                  (names (map (chain <>
+                                     (sxml-select <> (sxml-e? 's3:Name))
+                                     (cadar <>)
+                                     (make-bucket self <> (s3-client-region self)))
+                              buckets)))
+           ; buckets is #f if none are returned
+           (request-close req)
+           names))))
+
+;; NOTE: all bucket operations need the correct region for the bucket or they will 400
+(defmethod {create-bucket! s3-client}
+  (lambda (self bucket)
+    (using (self :- s3-client)
+      (let (req (s3-request/error self verb: 'PUT bucket: bucket))
+        (request-close req)
+        (void)))))
+
+; Gets a bucket struct that can be used to fetch objects.
+(defmethod {get-bucket s3-client} ; => bucket
+  (lambda (self bucket-name)
+    (using (self :- s3-client)
+      (if (s3-client::bucket-exists? self bucket-name)
+        (make-bucket self bucket-name self.region)
+        #f))))
```

Comment:
shouldn't the then/else clauses be inverted? make the bucket if it doesn't exist, as it is it will fail to create a new bucket.

Reply:
Also, let's wrap the returned bucket with its interface.

Reply:
I don't *think* this is inverted, the `s3client::bucket-exists?` checks that the bucket exists on the *server* side before returning an instance of the bucket struct. We don't want to be able to create a `bucket` instance that doesn't have a corresponding server representation.

Reply:
Oh I see. Yes, thats fine.

Maybe raise an exception instead of returning #f?